src/Flow.cpp

```diff
@@ -8745,19 +8773,39 @@ MinorConnectionStates Flow::calculateConnectionState(bool is_cumulative) {
       !(isTCPFlagSet(dst2src_tcp_flags,TH_SYN))) {
     if (!(isTCPFlagSet(dst2src_tcp_flags,TH_RST))) {
       if ((isTCPFlagSet(src2dst_tcp_flags,TH_RST))) {
+        updateSynFloodAlerts(false);
         return(setCurrentConnectionState(RSTOS0));
       } else {
         if ((isTCPFlagSet(src2dst_tcp_flags,TH_FIN))) {
+          updateSynFloodAlerts(false);
           return(setCurrentConnectionState(SH));
         } else {
-          return(setCurrentConnectionState(S0));
+            /* We are assuming that the TCP connection is established on
+             * server side when it sees the clients SYN
+             * this helps us detect SYN flood victims even if all the SYN
+             * packets are rejected or dropped*/
+            updateSynFloodAlerts(true);
+            return(setCurrentConnectionState(S0));
         }
       }
     } else {
+      updateSynFloodAlerts(false);
       return(setCurrentConnectionState(REJ));
     }
   }
 
+  /* Account for connection being closed after SYNACK*/
+  if ((isTCPFlagSet(src2dst_tcp_flags,TH_SYN)) &&
+      (isTCPFlagSet(dst2src_tcp_flags,TH_SYN)) &&
+      (isTCPFlagSet(dst2src_tcp_flags,TH_ACK)) &&
+      !(isTCPFlagSet(src2dst_tcp_flags,TH_ACK))&&
+      ((isTCPFlagSet(dst2src_tcp_flags,TH_RST))|| 
+      (isTCPFlagSet(src2dst_tcp_flags,TH_RST)) ||
+      (isTCPFlagSet(src2dst_tcp_flags,TH_FIN)) ||
+      (isTCPFlagSet(dst2src_tcp_flags,TH_FIN)))) {
+      updateSynFloodAlerts(false);
+  }
+
```

Comment:
The code wirtten in this way it's not very clean and hard to understand in the future when debugging. It would have been cleaner simply running a switch case at the end of the calculateConnectionState() function taking account for all the possible connection states and based on the value simply increase or decrease the counter.

Reply:
That makes sense! However, I am  a bit confused by the states we have defined as I feel they do not cover all possible connection states, such as when a connection is cancelled after a the server sends the SYNACK packet. Am I missing something or would it be ok to define another connection state?